dirvish-runall.8
.\"       $Id: dirvish-runall.8,v 12.0 2004/02/25 02:42:14 jw Exp $  $Name: Dirvish-1_2 $
.ds d \-\^\-
.ds o \fR[\fP
.ds c \fR]\fP
.ds | \fR|\fP
.ds bank \fIbank\fP
.ds vault \fIvault\fP
.ds branch \fIbranch\fP
.ds image \fIimage\fP
.de D
\\.B \*d\\$1
..
.de DR
\\.BR \*d\\$1 \\$2
..
.de Bi
\\.BI \\$1 " \\$2"
..
.de DI
\\.BI \*d\\$1 \\$2
..
.de Di
\\.BI \*d\\$1 " \\$2"
..
.de See
See \fB\\$1\fP for more details.
..
.de SeeIn
See \fB\\$1\fP in \fB\\$2\fP for more details.
..
.de multiple
Multiple \fB\\$1:\fP values will accumulate.
..
.de default
Default value: \fB\\$1\fP
..
.TH DIRVISH-RUNALL 8
.SH NAME
dirvish\-runall \- run a set of dirvish backup jobs.
.SH SYNOPSIS
.BI dirvish\-runall [OPTIONS]
.SH DESCRIPTION
Run all the jobs listed in the
.B Runall
field of the dirvish master configuration file.

Jobs are started immediately when
.B dirvish\-runall
is invoked; scheduling is handled by an external mechanism such as cron.

If
.B concurrent:
is set in the master configuration file,
up to that many
.B dirvish
processes will run in parallel.
The default is to run jobs sequentially.
.SH OPTIONS
Each option may be unambiguously abbreviated.
.TP
.Di config configfile
Read
.I configfile
for the
.B Runall:
field.
By default
.B dirvish\-runall
will use the
.B Runall:
field from the master configuration file.

This will only be used by
.B dirvish-runall
itself
and not be propagated to the
.B dirvish
processes.

.See dirvish.conf(5)
.TP
.D no\-run
.TP
.D dry\-run
Don't actually do anything.
Just display what would have happened.
.TP
.D quiet
Run quietly, only report errors.

Normally
.B dirvish\-runall
will report the dirvish commands invoked with timestamps.
.TP
.D version
Print the version information and exit.
.SH EXIT CODES
To facilitate further automation and integration of
.B dirvish-runall
with other tools
.B dirvish-runall
provides rationalised exit codes.
The exit codes are range based.  While the code for
a specific error may change from one version to another it
will remain within the specified range.  So don't test for
specific exit codes but instead test for a range of values.
To the degree possible higher value ranges indicate more
severe errors.
.TP
0
success
.TP
1-199
Number of
.B dirvish
jobs that failed.  If more than 199 jobs
failed 199 will be the exit code.
.TP
200-219
An error was encountered in loading a configuration file.
.TP
220-254
An error was detected in the configuration.
.TP
255
Incorrect usage.
.SH FILES
.TP
.B /etc/dirvish/master.conf
alternate master configuration file.
.TP
.B /etc/dirvish.conf
master configuration file.
.SH SEE ALSO
.nf
dirvish.conf(5)
.SH BUGS


dirvish-runall.pl

#       $Id: dirvish-runall.pl,v 12.0 2004/02/25 02:42:14 jw Exp $  $Name: Dirvish-1_2 $

$VERSION = ('$Name: Dirvish-1_2_1 $' =~ /Dirvish/i)
	? ('$Name: Dirvish-1_2_1 $' =~ m/^.*:\s+dirvish-(.*)\s*\$$/i)[0]
	: '1.1.2 patch' . ('$Id: dirvish-runall.pl,v 12.0 2004/02/25 02:42:14 jw Exp $'
		=~ m/^.*,v(.*:\d\d)\s.*$/)[0];
$VERSION =~ s/_/./g;


#########################################################################
#                                                         		#
#	Copyright 2002 and $Date: 2004/02/25 02:42:14 $
#                         Pegasystems Technologies and J.W. Schultz 	#
#                                                         		#
#	Licensed under the Open Software License version 2.0		#
#                                                         		#
#	This program is free software; you can redistribute it		#
#	and/or modify it under the terms of the Open Software		#
#	License, version 2.0 by Lauwrence E. Rosen.			#
#                                                         		#
#	This program is distributed in the hope that it will be		#
#	useful, but WITHOUT ANY WARRANTY; without even the implied	#
#	warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR		#
#	PURPOSE.  See the Open Software License for details.		#
#                                                         		#
#########################################################################

use Time::ParseDate;
use POSIX qw(strftime);
use POSIX ":sys_wait_h";
use Getopt::Long;

sub loadconfig;
sub seppuku;

sub usage
{
	my $message = shift(@_);

	length($message) and print STDERR $message, "\n\n";

	$! and exit(255); # because getopt seems to send us here for death

	print STDERR <<EOUSAGE;
USAGE
	dirvish-runall OPTIONS
OPTIONS
	--config configfile
	--no-run
	--quiet
	--version
EOUSAGE

	exit 255;
}

$Options = {
	version		=> sub {
			print STDERR "dirvish version $VERSION\n";
			exit(0);
		},
	help		=> \&usage,
};

GetOptions($Options, qw(
		config=s
		quiet no-run|dry-run
		version help
	)) or usage;

if ($$Options{config})
{
	$Config = loadconfig(undef, $$Options{config})
}
elsif ($CONFDIR =~ /dirvish$/ && -f "$CONFDIR.conf")
{
	$Config = loadconfig(undef, "$CONFDIR.conf");
}
elsif (-f "$CONFDIR/master.conf")
{
	$Config = loadconfig(undef, "$CONFDIR/master.conf");
}
elsif (-f "$CONFDIR/dirvish.conf")
{
	seppuku 250, <<EOERR;
ERROR: no master configuration file.
	An old $CONFDIR/dirvish.conf file found.
	Please read the dirvish release notes.
EOERR
}
else
{
	seppuku 251, "ERROR: no global configuration file";
}
$$Config{Dirvish} ||= '/usr/sbin/dirvish';

$$Config{concurrent} ||= '1';
@pidList = ();

$$Options{'no-run'} and $$Options{quiet} = 0;

$errors = 0;

for $sched (@{$$Config{Runall}})
{
	# Wait until we have a free slot
	until (scalar(@pidList) < int($$Config{concurrent})) {
		@pidList2 = ();
		for $pid (@pidList) {
			$wait_result = waitpid($pid, WNOHANG);
			if ($wait_result > 0) {
				# Increment the error counter if dirvish failed
				$? < 0 || $? >> 8 and ++$errors;
			} else {
				# Keep PID
				push(@pidList2, $pid);
			}
		}
		# Put all unsuccessful PIDs back in pidList
		@pidList = @pidList2;
		sleep(1);
	}

	# Spawn Child
	($vault, $itime) = split(/\s+/, $sched);
	$cmd = "$$Config{Dirvish} --vault $vault";
	$itime and $cmd .= qq[ --image-time "$itime"];
	$$Options{quiet}
		or printf "%s %s\n", strftime('%H:%M:%S', localtime), $cmd;
	$$Options{'no-run'} and next;
	$pid = fork();
	if ($pid > 0) {
		# Parent thread, register child pid
		push(@pidList, $pid);
	} elsif ($pid == -1) {
		# Failed fork
		print "Fork failed!\n";
		++$errors;
	} else {
		# pid == 0, child thread
		exec($cmd) || die "child exit";
	}
}

# Wait for remaining dirvish processes
for $pid (@pidList) {
	# Wait forever this time
	$wait_result = waitpid($pid, 0);
	if ($wait_result == -1 or $wait_result == $pid) {
		# Increment the error counter if dirvish failed
		$? < 0 || $? >> 8 and ++$errors;
	}
}

$$Options{quiet}
	or printf "%s %s\n", strftime('%H:%M:%S', localtime), 'done';


exit ($errors < 200 ? $errors : 199);
